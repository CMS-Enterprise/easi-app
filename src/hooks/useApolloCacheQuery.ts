// useApolloCacheQuery.ts
import {
  DocumentNode,
  OperationVariables,
  QueryHookOptions,
  QueryResult,
  TypedDocumentNode,
  useApolloClient,
  useQuery
} from '@apollo/client';

/** Custom hook to query the cache to determine if query result exists already */
export default function useApolloCacheQuery<
  TData = any,
  TVariables = OperationVariables
>(
  query: DocumentNode | TypedDocumentNode<TData, TVariables>,
  options?: QueryHookOptions<TData, TVariables>
): QueryResult<TData, TVariables> {
  const { cache } = useApolloClient();

  const queryResult = useQuery<TData, TVariables>(query, options);

  return {
    ...queryResult,
    loading:
      !cache.diff({
        query: cache.transformDocument(query),
        variables: options?.variables,
        returnPartialData: true,
        optimistic: false
      }).complete && queryResult.loading
  };
}
