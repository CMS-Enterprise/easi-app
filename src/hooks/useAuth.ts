import { useSelector } from 'react-redux';
import { useLazyQuery } from '@apollo/client';
import { useFlags } from 'launchdarkly-react-client-sdk';

import {
  BASIC_USER_PROD,
  GOVTEAM_DEV,
  GOVTEAM_PROD,
  TRB_ADMIN_DEV,
  TRB_ADMIN_PROD
} from 'constants/jobCodes';
import GetSystemIntakeGrbReviewersQuery from 'queries/GetSystemIntakeGrbReviewersQuery';
import {
  GetSystemIntakeGrbReviewers,
  GetSystemIntakeGrbReviewersVariables
} from 'queries/types/GetSystemIntakeGrbReviewers';
import { AppState } from 'reducers/rootReducer';

type UseAuthReturn = {
  /**
   * Object with helper functions to check job codes / roles
   *
   * Returns undefined if user is not set
   */
  user:
    | {
        isTrbAdmin: () => boolean;
        isGrtReviewer: () => boolean;
        isBasicUser: () => boolean;
        isGrbReviewer: (systemIntakeId: string) => boolean;
      }
    | undefined;
  /** Returns true if function with query is called and loading */
  loading: boolean;
};

/** Hook that returns functions to check for roles / job codes */
const useAuth = (): UseAuthReturn => {
  const auth = useSelector((state: AppState) => state.auth);
  const { groups, euaId } = auth;

  const flags = useFlags();

  const [getGrbReviewers, { data, called, loading }] = useLazyQuery<
    GetSystemIntakeGrbReviewers,
    GetSystemIntakeGrbReviewersVariables
  >(GetSystemIntakeGrbReviewersQuery);

  const isUserSet = auth.isUserSet && !loading;

  /** Check if user is TRB admin */
  const isTrbAdmin = (): boolean => {
    if (flags.downgradeTrbAdmin) {
      return false;
    }

    if (groups.includes(TRB_ADMIN_DEV) || groups.includes(TRB_ADMIN_PROD)) {
      return true;
    }

    return false;
  };

  /** Check if user is GrtReviewer */
  const isGrtReviewer = (): boolean => {
    if (flags.downgradeGovTeam) {
      return false;
    }

    if (groups.includes(GOVTEAM_DEV) || groups.includes(GOVTEAM_PROD)) {
      return true;
    }

    return false;
  };

  /** Check if basic user */
  const isBasicUser = (): boolean => {
    if (groups.includes(BASIC_USER_PROD)) {
      return true;
    }
    if (groups.length === 0) {
      return true;
    }
    if (!isGrtReviewer() && !isTrbAdmin()) {
      return true;
    }
    return false;
  };

  /** Check if user is set as system intake GRB reviewer */
  const isGrbReviewer = (systemIntakeId: string) => {
    if (!called) {
      getGrbReviewers({
        variables: { id: systemIntakeId }
      });
    }

    const grbReviewers = data?.systemIntake?.grbReviewers;

    if (!grbReviewers || grbReviewers.length === 0) {
      return false;
    }

    return grbReviewers.some(
      ({ userAccount }) => userAccount.username === euaId
    );
  };

  const user: UseAuthReturn['user'] = {
    isTrbAdmin,
    isGrtReviewer,
    isBasicUser,
    isGrbReviewer
  };

  return {
    user: isUserSet || !loading ? user : undefined,
    loading
  };
};

export default useAuth;
