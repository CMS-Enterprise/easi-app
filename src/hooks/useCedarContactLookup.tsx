/** Custom hook for retrieving contacts from Cedar by common name */

import { useMemo, useState } from 'react';
import { ApolloQueryResult, useQuery } from '@apollo/client';

import GetCedarContactsQuery from 'queries/GetCedarContactsQuery';
import { GetCedarContacts } from 'queries/types/GetCedarContacts';
import { CedarContactProps } from 'types/systemIntake';

function useCedarContactLookup(): [
  CedarContactProps[],
  (commonName: string) => void
];

function useCedarContactLookup(
  query: string,
  euaUserId: string
): CedarContactProps | undefined;

function useCedarContactLookup(
  query?: string,
  euaUserId?: string
):
  | [CedarContactProps[], (commonName: string) => void]
  | CedarContactProps
  | undefined {
  const { data, refetch } = useQuery<GetCedarContacts>(GetCedarContactsQuery, {
    variables: { commonName: query },
    skip: !query
  });
  const [contacts, setContacts] = useState<CedarContactProps[]>(
    data?.cedarPersonsByCommonName || []
  );

  const contactByEuaUserId = useMemo<CedarContactProps | undefined>(() => {
    return data?.cedarPersonsByCommonName.find(
      contact => contact.euaUserId === euaUserId
    );
  }, [data?.cedarPersonsByCommonName, euaUserId]);

  const queryCedarContacts = (commonName: string) => {
    // Removes eua id from "John Doe, ABCD" format when querying
    const commonNameQuery = commonName.trim().split(',')[0];

    if (commonNameQuery.length > 1) {
      refetch({ commonName: commonNameQuery })
        .then((response: ApolloQueryResult<GetCedarContacts>) => {
          setContacts(
            sortCedarContacts(
              response?.data?.cedarPersonsByCommonName,
              commonNameQuery
            )
          );
        })
        .catch(() => {
          // If error, set contacts to empty array
          setContacts([]);
        });
    }
  };

  return euaUserId ? contactByEuaUserId : [contacts, queryCedarContacts];
}

const sortCedarContacts = (
  contacts: CedarContactProps[],
  query: string
): CedarContactProps[] => {
  return [...contacts].sort((a, b) => {
    const result =
      a.commonName.toLowerCase().search(query) -
      b.commonName.toLowerCase().search(query);
    if (result > 0) return 1;
    if (result < 0) return -1;
    return 0;
  });
};

export default useCedarContactLookup;
