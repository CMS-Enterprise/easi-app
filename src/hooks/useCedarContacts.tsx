import { useCallback, useMemo } from 'react';
import { useQuery } from '@apollo/client';

import GetCedarContactsQuery from 'queries/GetCedarContactsQuery';
import { GetCedarContacts } from 'queries/types/GetCedarContacts';
import { CedarContactProps } from 'types/systemIntake';

export default function useCedarContacts(): {
  contacts: CedarContactProps[] | null;
  getContactByEua: (euaUserId: string) => CedarContactProps | null;
} {
  const { data } = useQuery<GetCedarContacts>(GetCedarContactsQuery, {
    variables: { commonName: '' }
  });
  const contacts = useMemo(() => data?.cedarPersonsByCommonName || null, [
    data?.cedarPersonsByCommonName
  ]);
  const getContactByEua = useCallback(
    (euaUserId: string) => {
      return (
        (contacts || []).find(contact => contact.euaUserId === euaUserId) ??
        null
      );
    },
    [contacts]
  );
  return { contacts, getContactByEua };
}
