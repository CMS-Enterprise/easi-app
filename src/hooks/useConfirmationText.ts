import { useEffect, useState } from 'react';
import { useHistory, useLocation } from 'react-router-dom';

// useConfirmationText is a hook used to get the confirmation text
// passed through react router state.
// const confirmationText = useConfirmationText();
//
// Two common ways to pass state through react-router (there are more)
// history.push('/some/url', { confirmationText: 'Confirmed' });
// or
// const location = { pathname: '/some/url', state: { confirmationText: 'Confirmed' }};
// <Link to={location} />
function useConfirmationText() {
  const location = useLocation<any>();
  const history = useHistory();
  const [confirmationText, setConfirmationText] = useState('');

  useEffect(() => {
    if (location.state && location.state.confirmationText) {
      setConfirmationText(location.state.confirmationText);
      history.replace({
        state: {}
      });
    }

    // unset confirmationText if it's null
    if (!location.state?.confirmationText) {
      setConfirmationText('');
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [history.location.pathname]);

  return confirmationText;
}

export default useConfirmationText;
