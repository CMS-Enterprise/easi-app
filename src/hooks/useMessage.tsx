import React, {
  createContext,
  ReactNode,
  useContext,
  useEffect,
  useState
} from 'react';
import { useLocation } from 'react-router-dom';

import Alert, { AlertProps } from 'components/shared/Alert';

type MessageProps = {
  message?: string | React.ReactElement;
  alertProps?: Omit<AlertProps, 'children'>;
};

type SetMessage = (
  message?: string | React.ReactElement,
  alertProps?: Omit<AlertProps, 'children'>
) => void;

const MessageContext = createContext<
  | {
      Message: ({ className }: { className?: string }) => JSX.Element | null;
      showMessage: SetMessage;
      showMessageOnNextPage: SetMessage;
    }
  | undefined
>(undefined);

// MessageProvider manages the state necessary for child components to
// use the useMessage hook.
const MessageProvider = ({ children }: { children: ReactNode }) => {
  const [queuedMessage, setQueuedMessage] = useState<MessageProps>({});
  const [message, setMessage] = useState<MessageProps>({});

  const location = useLocation();

  const [lastPathname, setLastPathname] = useState(location.pathname);

  const Message = (props: { className?: string }) => {
    // If no message, return null
    if (!message.message) return null;

    // If message has alert props, wrap in component
    if (message.alertProps) {
      return (
        <Alert {...message.alertProps} {...props}>
          {message.message}
        </Alert>
      );
    }

    // Return plain message
    return <>{message.message}</>;
  };

  useEffect(() => {
    if (lastPathname !== location.pathname) {
      setMessage(queuedMessage);
      setQueuedMessage({});
      setLastPathname(location.pathname);
    }
  }, [message, queuedMessage, lastPathname, location.pathname]);

  return (
    <MessageContext.Provider
      value={{
        Message,
        showMessage: (value, alertProps) =>
          setMessage({
            message: value,
            alertProps
          }),
        showMessageOnNextPage: (value, alertProps) =>
          setQueuedMessage({
            message: value,
            alertProps
          })
      }}
    >
      {children}
    </MessageContext.Provider>
  );
};

// useMessage is a hook that provides the ability to show notifications to the
// user. Components will use the hook to get access to two methods:
//
// showMessage: display a message to the user immediately. This is useful
// when an interaction won't cause an immediate redirect to a new page.
//
// showMessageOnNextPage: schedule a message to be shown on the next
// visited by the user. Pages are identified using their URL and therefore
// messages won't appear until the URL changes.
//
// Currently there is only support for showing a single message.
const useMessage = () => {
  const context = useContext(MessageContext);
  if (context === undefined) {
    throw new Error('useMessage must be used within a MessageProvider');
  }

  return context;
};

export default useMessage;
export { useMessage, MessageProvider };
