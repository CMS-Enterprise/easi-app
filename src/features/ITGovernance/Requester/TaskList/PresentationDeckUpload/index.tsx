import React from 'react';
import { Controller } from 'react-hook-form';
import { useTranslation } from 'react-i18next';
import { useHistory, useParams } from 'react-router-dom';
import { FileInput, Form, FormGroup, Icon } from '@trussworks/react-uswds';
import Pager from 'features/TechnicalAssistance/Requester/RequestForm/Pager';
import {
  SystemIntakeGRBPresentationLinksInput,
  useSetSystemIntakeGRBPresentationLinksMutation,
  useUploadSystemIntakeGRBPresentationDeckMutation
} from 'gql/generated/graphql';

import { Alert } from 'components/Alert';
import { useEasiForm } from 'components/EasiForm';
import IconLink from 'components/IconLink';
import Label from 'components/Label';
import useMessage from 'hooks/useMessage';
import { ITGovernanceViewType } from 'types/itGov';
import { fileToBase64File } from 'utils/downloadFile';

type PresentationLinkFields = Omit<
  SystemIntakeGRBPresentationLinksInput,
  | 'systemIntakeID'
  | 'recordingLink'
  | 'recordingPasscode'
  | 'transcriptLink'
  | 'transcriptFileName'
>;

type UploadFormProps = {
  type: ITGovernanceViewType;
};

/**
 * Presentation deck document upload form
 */
const PresentationDeckUpload = ({ type = 'requester' }: UploadFormProps) => {
  const { t } = useTranslation('grbReview');

  const history = useHistory();

  const { systemId } = useParams<{
    systemId: string;
  }>();

  const { showMessageOnNextPage, showMessage } = useMessage();

  // TODO: May need to incorporate new mutation for requester, once BE work is complete
  // Form would need to toggle between two mutations based on user type
  const [setPresentationLinks] = useSetSystemIntakeGRBPresentationLinksMutation(
    { refetchQueries: ['GetSystemIntake'] }
  );
  const [upload] = useUploadSystemIntakeGRBPresentationDeckMutation({
    refetchQueries: ['GetSystemIntake']
  });

  const {
    handleSubmit,
    control,
    watch,
    formState: { errors, isSubmitting }
  } = useEasiForm<PresentationLinkFields>({
    defaultValues: {
      presentationDeckFileData: {
        name: ''
      } as File
    }
  });

  const requestDetailsLink =
    type === 'requester'
      ? `/governance-task-list/${systemId}`
      : `/it-governance/${systemId}/grb-review`;

  const submit = handleSubmit(async values => {
    const presentationDeckFileData = values.presentationDeckFileData?.size
      ? await fileToBase64File(values.presentationDeckFileData)
      : undefined;

    if (type === 'requester') {
      upload({
        variables: {
          input: {
            systemIntakeID: systemId,
            presentationDeckFileData
          }
        }
      })
        .then(() => {
          showMessageOnNextPage(
            t('presentationLinks.presentationUpload.success'),
            {
              type: 'success'
            }
          );
          history.push(requestDetailsLink);
        })
        .catch(() => {
          showMessage(t('presentationLinks.presentationUpload.error'), {
            type: 'error',
            className: 'margin-top-4'
          });
        });
    } else {
      setPresentationLinks({
        variables: {
          input: {
            systemIntakeID: systemId,
            presentationDeckFileData
          }
        }
      })
        .then(() => {
          showMessageOnNextPage(
            t('presentationLinks.presentationUpload.success'),
            {
              type: 'success'
            }
          );
          history.push(requestDetailsLink);
        })
        .catch(() => {
          showMessage(t('presentationLinks.presentationUpload.error'), {
            type: 'error',
            className: 'margin-top-4'
          });
        });
    }
  });

  return (
    <>
      <div className="tablet:grid-col-12 desktop:grid-col-8 margin-top-6 margin-bottom-8 padding-bottom-4">
        <h1 className="margin-bottom-1">
          {t('presentationLinks.presentationUpload.header')}
        </h1>

        <p className="margin-top-1 margin-bottom-2 font-body-md line-height-body-5 text-light">
          {t('presentationLinks.presentationUpload.description')}
        </p>

        <IconLink to={requestDetailsLink} icon={<Icon.ArrowBack />}>
          {type === 'requester'
            ? t('presentationLinks.presentationUpload.dontUploadRequester')
            : t('presentationLinks.presentationUpload.dontUploadAdmin')}
        </IconLink>

        <Form className="maxw-full" onSubmit={submit}>
          {/* Upload field */}
          <FormGroup
            error={!!errors.presentationDeckFileData}
            className="margin-top-5"
          >
            <Label htmlFor="fileData">
              {t('presentationLinks.presentationUpload.selectFile')}
            </Label>

            <p className="usa-hint margin-bottom-0 margin-top-05">
              {t('presentationLinks.presentationUpload.recommendedFileTypes')}
            </p>

            <Controller
              control={control}
              name="presentationDeckFileData"
              rules={{
                required: true
              }}
              render={({ field: { value, ...field } }) => (
                <FileInput
                  {...field}
                  id={field.name}
                  name={field.name}
                  onChange={e => {
                    field.onChange(e.currentTarget?.files?.[0]);
                  }}
                  accept=".pdf,.doc,.docx,.ppt,.pptx,.xls,.xlsx"
                />
              )}
            />
          </FormGroup>

          <Alert type="info" slim className="margin-top-5">
            {t('presentationLinks.uploadAlert')}
          </Alert>

          <Pager
            next={{
              text: t('presentationLinks.presentationUpload.upload'),
              disabled: !watch('presentationDeckFileData')?.name || isSubmitting
            }}
            taskListUrl={requestDetailsLink}
            saveExitText={
              type === 'requester'
                ? t('presentationLinks.presentationUpload.dontUploadRequester')
                : t('presentationLinks.presentationUpload.dontUploadAdmin')
            }
            submitDisabled
            border={false}
            className="margin-top-4"
          />
        </Form>
      </div>
    </>
  );
};

export default PresentationDeckUpload;
