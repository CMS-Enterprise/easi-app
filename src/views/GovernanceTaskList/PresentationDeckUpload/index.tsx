import React from 'react';
import { Controller } from 'react-hook-form';
import { Trans, useTranslation } from 'react-i18next';
import { useHistory, useParams } from 'react-router-dom';
import { useMutation } from '@apollo/client';
import { yupResolver } from '@hookform/resolvers/yup';
import { FileInput, Form, FormGroup, Icon } from '@trussworks/react-uswds';

import { useEasiForm } from 'components/EasiForm';
import { Alert } from 'components/shared/Alert';
import IconLink from 'components/shared/IconLink';
import Label from 'components/shared/Label';
import useMessage from 'hooks/useMessage';
import GetSystemIntakeQuery from 'queries/GetSystemIntakeQuery';
import { CreateSystemIntakeDocumentQuery } from 'queries/SystemIntakeDocumentQueries';
import {
  CreateSystemIntakeDocument,
  CreateSystemIntakeDocumentVariables
} from 'queries/types/CreateSystemIntakeDocument';
import { CreateSystemIntakeDocumentInput } from 'types/graphql-global-types';
import { fileToBase64File } from 'utils/downloadFile';
import { documentSchema } from 'validations/systemIntakeSchema';
import Pager from 'views/TechnicalAssistance/RequestForm/Pager';

type DocumentUploadFields = Omit<CreateSystemIntakeDocumentInput, 'requestID'>;

type UploadFormProps = {
  type: 'admin' | 'requester';
};

/**
 * Presentation deck document upload form
 */
const PresentationDeckUpload = ({ type = 'requester' }: UploadFormProps) => {
  const { t } = useTranslation('grbReview');

  const history = useHistory();

  const { systemId } = useParams<{
    systemId: string;
  }>();

  const { Message, showMessageOnNextPage, showMessage } = useMessage();

  const [createDocument] = useMutation<
    CreateSystemIntakeDocument,
    CreateSystemIntakeDocumentVariables
  >(CreateSystemIntakeDocumentQuery, {
    refetchQueries: [
      {
        query: GetSystemIntakeQuery,
        variables: {
          id: systemId
        }
      }
    ]
  });

  const {
    control,
    register,
    watch,
    handleSubmit,
    formState: { isSubmitting, errors, isValid }
  } = useEasiForm<DocumentUploadFields>({
    resolver: yupResolver(documentSchema),
    context: { type }
  });

  const requestDetailsLink =
    type === 'requester'
      ? `/governance-task-list/${systemId}`
      : `/it-governance/${systemId}/grb-review`;

  const submit = handleSubmit(async ({ otherTypeDescription, ...formData }) => {
    const newFile = await fileToBase64File(formData.fileData);
    createDocument({
      variables: {
        input: {
          fileData: newFile,
          version: formData.version,
          documentType: formData.documentType,
          // If type is set to 'Other', include description field
          ...(formData.documentType === 'OTHER'
            ? { otherTypeDescription }
            : {}),
          requestID: systemId,
          sendNotification: formData.sendNotification
        }
      }
    })
      .then(() => {
        showMessageOnNextPage(
          t('presentationLinks.presentationUpload.success'),
          {
            type: 'success'
          }
        );
        history.push(requestDetailsLink);
      })
      .catch(() => {
        showMessage(t('presentationLinks.presentationUpload.error'), {
          type: 'error',
          className: 'margin-top-4'
        });
      });
  });

  return (
    <>
      <Message />

      <div className="tablet:grid-col-12 desktop:grid-col-8 margin-top-6 margin-bottom-8 padding-bottom-4">
        <h1 className="margin-bottom-1">
          {t('presentationLinks.presentationUpload.header')}
        </h1>

        <p className="margin-top-1 margin-bottom-2 font-body-md line-height-body-5 text-light">
          {t('presentationLinks.presentationUpload.description')}
        </p>

        <IconLink to={requestDetailsLink} icon={<Icon.ArrowBack />}>
          {type === 'requester'
            ? t('presentationLinks.presentationUpload.dontUploadRequester')
            : t('presentationLinks.presentationUpload.dontUploadAdmin')}
        </IconLink>

        <Form className="maxw-full" onSubmit={submit}>
          {/* Upload field */}
          <FormGroup error={!!errors.fileData} className="margin-top-5">
            <Label htmlFor="fileData">
              {t('presentationLinks.presentationUpload.selectFile')}
            </Label>

            <p className="usa-hint margin-bottom-0 margin-top-05">
              {t('presentationLinks.presentationUpload.recommendedFileTypes')}
            </p>

            <Controller
              control={control}
              name="fileData"
              render={({ field: { value, ...field } }) => (
                <FileInput
                  {...field}
                  id={field.name}
                  name={field.name}
                  onChange={e => {
                    field.onChange(e.currentTarget?.files?.[0]);
                  }}
                  accept=".pdf,.ppt,.pptx"
                />
              )}
            />
          </FormGroup>

          <Alert type="info" slim className="margin-top-5">
            {t('presentationLinks.uploadAlert')}
          </Alert>

          <Pager
            next={{
              text: t('presentationLinks.presentationUpload.upload'),
              disabled: !isValid || isSubmitting
            }}
            taskListUrl={requestDetailsLink}
            saveExitText={
              type === 'requester'
                ? t('presentationLinks.presentationUpload.dontUploadRequester')
                : t('presentationLinks.presentationUpload.dontUploadAdmin')
            }
            border={false}
            className="margin-top-4"
          />
        </Form>
      </div>
    </>
  );
};

export default PresentationDeckUpload;
