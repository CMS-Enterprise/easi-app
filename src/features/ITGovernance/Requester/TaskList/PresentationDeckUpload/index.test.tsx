import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { MockedProvider } from '@apollo/client/testing';
import { render, screen } from '@testing-library/react';
import i18next from 'i18next';

import { MessageProvider } from 'hooks/useMessage';

import PresentationDeckUpload from '.';

describe('UploadForm page', () => {
  it('renders correct navigation text if admin', async () => {
    render(
      <MemoryRouter>
        <MockedProvider>
          <MessageProvider>
            <PresentationDeckUpload type="admin" />
          </MessageProvider>
        </MockedProvider>
      </MemoryRouter>
    );

    expect(
      screen.getAllByText(
        i18next.t<string>(
          'grbReview:presentationLinks.presentationUpload:dontUploadAdmin'
        )
      )[0]
    );
  });

  it('renders correct navigation text if requester and takes a snapshot', async () => {
    const { asFragment } = render(
      <MemoryRouter>
        <MockedProvider>
          <MessageProvider>
            <PresentationDeckUpload type="requester" />
          </MessageProvider>
        </MockedProvider>
      </MemoryRouter>
    );

    expect(
      screen.getAllByText(
        i18next.t<string>(
          'grbReview:presentationLinks.presentationUpload:dontUploadRequester'
        )
      )[0]
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
