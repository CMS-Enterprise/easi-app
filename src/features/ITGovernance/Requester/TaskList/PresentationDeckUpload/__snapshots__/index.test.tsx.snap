// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UploadForm page > renders correct navigation text if requester and takes a snapshot 1`] = `
<DocumentFragment>
  <div
    class="tablet:grid-col-12 desktop:grid-col-8 margin-top-6 margin-bottom-8 padding-bottom-4"
  >
    <h1
      class="margin-bottom-1"
    >
      Upload GRB presentation
    </h1>
    <p
      class="margin-top-1 margin-bottom-2 font-body-md line-height-body-5 text-light"
    >
      Upload the presentation deck that this project will share with the Governance Review Board (GRB). You may change the uploaded file at any time before the GRB review period starts. 
    </p>
    <a
      class="usa-link display-flex flex-align-center"
      href="/governance-task-list/undefined"
    >
      <svg
        class="usa-icon margin-right-1"
        focusable="false"
        height="1em"
        role="img"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
        />
      </svg>
      Don’t upload and return to the governance task list
    </a>
    <form
      class="usa-form maxw-full"
      data-testid="form"
    >
      <div
        class="usa-form-group margin-top-5"
        data-testid="formGroup"
      >
        <label
          class="usa-label"
          for="fileData"
        >
          Select your file
        </label>
        <p
          class="usa-hint margin-bottom-0 margin-top-05"
        >
          Select a PDF, PPT, PPTX, DOC, DOCX, XLS, or XLSX
        </p>
        <div
          class="usa-file-input"
          data-testid="file-input"
        >
          <div
            class="usa-file-input__target"
            data-testid="file-input-droptarget"
          >
            <div
              aria-hidden="true"
              class="usa-file-input__instructions"
              data-testid="file-input-instructions"
            >
              <span
                class="usa-file-input__drag-text"
              >
                Drag file here or 
              </span>
              <span
                class="usa-file-input__choose"
              >
                choose from folder
              </span>
            </div>
            <div
              class="usa-file-input__box"
              data-testid="file-input-box"
            />
            <input
              accept=".pdf,.doc,.docx,.ppt,.pptx,.xls,.xlsx"
              class="usa-file-input__input"
              data-testid="file-input-input"
              id="presentationDeckFileData"
              name="presentationDeckFileData"
              type="file"
            />
          </div>
        </div>
      </div>
      <div
        class="usa-alert usa-alert--info usa-alert--slim margin-top-5"
        data-testid="alert"
      >
        <div
          class="usa-alert__body"
        >
          <p
            class="usa-alert__text"
          >
            To keep CMS safe, documents are scanned for viruses after uploading. If something goes wrong, we'll let you know.
          </p>
        </div>
      </div>
      <div
        class="margin-top-4"
      >
        <ul
          class="usa-button-group"
        >
          <li
            class="usa-button-group__item"
          >
            <div
              class="display-flex flex-align-center"
            >
              <button
                class="usa-button margin-top-0"
                data-testid="button"
                disabled=""
                type="submit"
              >
                Upload presentation
              </button>
            </div>
          </li>
        </ul>
        <button
          class="usa-button usa-button--unstyled margin-top-2 display-flex flex-align-center"
          data-testid="button"
          type="button"
        >
          <svg
            class="usa-icon margin-right-05"
            focusable="false"
            height="1em"
            role="img"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
            />
          </svg>
          Don’t upload and return to the governance task list
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
